src/main/java/com/amazonaws/encryptionsdk/internal/AesGcmJceKeyCipher.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except
+ * in compliance with the License. A copy of the License is located at
+ *
+ * http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package com.amazonaws.encryptionsdk.internal;
+
+import javax.crypto.Cipher;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.GCMParameterSpec;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.security.InvalidKeyException;
+import java.security.Key;
+import java.security.SecureRandom;
+import java.util.Map;
+
+/**
+ * A JceKeyCipher based on the Advanced Encryption Standard in Galois/Counter Mode.
+ */
+class AesGcmJceKeyCipher extends JceKeyCipher {
+    private static final int NONCE_LENGTH = 12;
+    private static final int TAG_LENGTH = 128;
+    private static final String TRANSFORMATION = "AES/GCM/NoPadding";
+    private final SecureRandom rnd = new SecureRandom();
+
+    AesGcmJceKeyCipher(SecretKey key) {
+        super(key, key);
+    }
+
+    private static byte[] specToBytes(final GCMParameterSpec spec) {
+        final byte[] nonce = spec.getIV();
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        try (final DataOutputStream dos = new DataOutputStream(baos)) {
+            dos.writeInt(spec.getTLen());
+            dos.writeInt(nonce.length);
+            dos.write(nonce);
+            dos.close();
+            baos.close();
+        } catch (final IOException ex) {
+            throw new AssertionError("Impossible exception", ex);
+        }
+        return baos.toByteArray();
+    }
+
+    private static GCMParameterSpec bytesToSpec(final byte[] data, final int offset) throws InvalidKeyException {
+        final ByteArrayInputStream bais = new ByteArrayInputStream(data, offset, data.length - offset);
+        try (final DataInputStream dis = new DataInputStream(bais)) {
+            final int tagLen = dis.readInt();
+            final int nonceLen = dis.readInt();
+
+            if (tagLen != TAG_LENGTH) {
+                throw new InvalidKeyException(String.format("Authentication tag length must be %s", TAG_LENGTH));
+            }
+
+            if (nonceLen != NONCE_LENGTH) {
+                throw new InvalidKeyException(String.format("Initialization vector (IV) length must be %s", NONCE_LENGTH));
+            }
```

Comment:
This is the only part of this PR that isn't a refactor. I made this change to bring this in line with the RawAesKeyring spec